pcap-options.c

```diff
@@ -0,0 +1,141 @@
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <sys/types.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "pcap-int.h"
+
+/*
+ * Private data for storing sets of options.
+ * This is to be used by functions which seem to grow additional options
+ * This is avoid ABI explosition of do_thing, and do_thing_with_opt_bar(),
+ * and then do_thing_with_opt_bar_baz(...). Instead a "pcap_option" should be
+ * created to which get/set shall be done.
+ *
+ * each option shall have an element in enum pcap_option_name {}.
+ */
+struct pcap_options {
+        int   tstamp_precision;
+        const char *io_read_plugin;
+        const char *io_write_plugin;
+};
+
+pcap_options *pcap_alloc_option(void)
+{
+        pcap_options *po = malloc(sizeof(struct pcap_options));
+        memset(po, 0, sizeof(struct pcap_options));
+        return po;  // caller has to check for NULL anyway.
+}
+
+void pcap_free_option(pcap_options *po)
+{
+        if(po != NULL) {
+                if(po->io_read_plugin) free((void *)po->io_read_plugin);
+                if(po->io_write_plugin) free((void *)po->io_write_plugin);
+                free((void *)po);
+        }
+}
+
+/* Return 0 on success, -1 on failure invalid option, -2 on type mis-match */
+int pcap_set_option_string(pcap_options *po,
+                           enum   pcap_option_name pon,
+                           const char *value)
+{
+        const char *saved = strdup(value);
+        switch(pon) {
+        case PON_TSTAMP_PRECISION:
+                free((void *)saved);
+                return -2;
+
+        case PON_IO_READ_PLUGIN:
+                po->io_read_plugin = saved;
+                break;
+        case PON_IO_WRITE_PLUGIN:
+                po->io_write_plugin= saved;
+                break;
+        default:
+                free((void *)saved);
+                return -1;
+        }
+        return 0;
+}
+
+/* Return 0 on success, -1 on failure invalid option, -2 on type mis-match */
+int pcap_set_option_int(pcap_options *po,
+                        enum   pcap_option_name pon,
+                        const int value)
+{
+        switch(pon) {
+        case PON_TSTAMP_PRECISION:
+                po->tstamp_precision = value;
+                break;
+
+        case PON_IO_READ_PLUGIN:
+        case PON_IO_WRITE_PLUGIN:
+                return -2;
+        default:
+                return -1;
+        }
+        return 0;
+}
+
+const char *pcap_get_option_string(pcap_options *po,
+                                   enum   pcap_option_name pon)
+{
+        switch(pon) {
+        case PON_TSTAMP_PRECISION:
+                return NULL;
+
+        case PON_IO_READ_PLUGIN:
+                return po->io_read_plugin;
+        case PON_IO_WRITE_PLUGIN:
+                return po->io_write_plugin;
+        default:
+                return NULL;
+        }
+        return NULL;
+}
+
+
+/* return int value, or zero for not-found, mis-type */
```

Comment:
You might want to limit the value to 0-to-INT_MAX, and have -1 mean "not found", or have a Boolean success/failure return and return the value through a pointer.

(Many's the time when I wished various programming languages had the notion of a function failing as an explicit concept - "failure *is* an option". :-)  Perhaps that's what exceptions are for, but C doesn't have them.)

Reply:
Yes, I was thinking about whether to return the value as a result parameter.
RUST and GO both seem to do failure is an option well.  Rust seems very good at it.
I've been immersed in Rust for 6 weeks.  

I will revise things in that way.  I just wasn't sure if it was worth it.